test/integration/crl_test.go

```diff
@@ -97,3 +101,84 @@ func TestCRLPipeline(t *testing.T) {
 	test.AssertEquals(t, string(reason), "5")
 	resp.Body.Close()
 }
+
+func TestTemporalAndExplicitShardingCoexist(t *testing.T) {
+	db, err := sql.Open("mysql", vars.DBConnSAIntegrationFullPerms)
+	if err != nil {
+		t.Fatalf("sql.Open: %s", err)
+	}
+	// Insert an old, revoked certificate in the certificateStatus table. Importantly this
+	// serial has the 7f prefix, which is in test/config-next/crl-updater.json in the
+	// `temporallyShardedPrefixes` list.
+	oldSerial := "7faa39be44fc95f3d19befe3cb715848e601"
+	issuerID := 43104258997432926
+	_, err = db.Exec(`DELETE FROM certificateStatus WHERE serial = ?`, oldSerial)
+	if err != nil {
+		t.Fatalf("deleting old certificateStatus row: %s", err)
+	}
+	_, err = db.Exec(fmt.Sprintf(`
+		INSERT INTO certificateStatus (serial, issuerID, notAfter, status, ocspLastUpdated, revokedDate, revokedReason, lastExpirationNagSent)
+		VALUES ("%s", %d, "%s", "revoked", NOW(), NOW(), 0, 0);`,
+		oldSerial, issuerID, time.Now().Add(24*time.Hour).Format("2006-01-02 15:04:05")))
+	if err != nil {
+		t.Fatalf("inserting old certificateStatus row: %s", err)
+	}
+
+	client, err := makeClient()
+	if err != nil {
+		t.Fatalf("creating acme client: %s", err)
+	}
+
+	certKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		t.Fatalf("creating cert key: %s", err)
+	}
+
+	result, err := authAndIssue(client, certKey, []string{random_domain()}, true)
+	if err != nil {
+		t.Fatalf("authAndIssue: %s", err)
+	}
+
+	cert := result.certs[0]
+	err = client.RevokeCertificate(
+		client.Account,
+		cert,
+		client.PrivateKey,
+		0,
+	)
+	if err != nil {
+		t.Fatalf("revoking: %s", err)
+	}
+
+	// Reset the "leasedUntil" column to prepare for another round of CRLs.
+	fc := clock.NewFake()
+	_, err = db.Exec(`UPDATE crlShards SET leasedUntil = ?`, fc.Now().Add(-time.Minute))
+	if err != nil {
+		t.Fatalf("resetting crlShards.leasedUntil: %s", err)
+	}
+
+	runUpdater(t, path.Join(os.Getenv("BOULDER_CONFIG_DIR"), "crl-updater.json"))
+
+	allCRLs := getAllCRLs(t)
+	allSeen := make(map[string]bool)
+	// Range over CRLs from all issuers, because the "old" certificate (7faa...) has a
+	// different issuer than the "new" certificate issued by `authAndIssue`, which
+	// has a random issuer.
+	for _, crls := range allCRLs {
+		seen := make(map[string]bool)
```

Comment:
You still don't need separate maps for within-an-issuer and across-issuers: a given serial only has a single issuer, so a single `seen` map will suffice.